demo_modules/chicago-brick-live/chicago-brick-live.js

```diff
@@ -0,0 +1,264 @@
+canvas.writeText(screen.width/2, screen.height/2, "${message}", "white", "80px Arial", {textAlign: "center"});
+canvas.draw.image(10, 10, "https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png", 0.5);`;
+}
+
+function getClientKey(client) {
+  return `${client.x},${client.y}`;
+}
+//
+// Server Module
+//
+class LiveClientServer extends ServerModuleInterface {
+  constructor(config) {
+    super();
+
+    this.config = _.defaults(config, DEFAULT_CONFIG);
+    debug(`Attempting to use codeserver at ${this.config.codeServer}`);
+
+    // All clients (x, y) that have ever connected.  Used to notify all clients
+    // of global code changes (e.g., codeserver comes online) and cache code.
+    this.clients = {};
+
+    const inst = this;
+    this.codeServer = require('socket.io-client')(this.config.codeServer);
+
+    // Setup connection to code server.
+    inst.codeServer.on('connect', function () {
+      debug(`Connected to code server (${inst.config.codeServer}).`);
+
+      // When code server connection is made re-request code for all clients
+      // we know about.
+      // #TODO Is there any client list provided by brick?
+      for (var key in inst.clients) {
+        inst.requestCode(inst.clients[key].client);
+      }
+    });
+
+    inst.codeServer.on('disconnect', function () {
+      debug('Disconnected from code server.');
+    });
+
+    inst.codeServer.on('code', function(data) {
+      // Make a unique key of the form 'x,y' so we can use a dictionary for clients.
+      var key = getClientKey(data.client);
+      debug(`Received new info for client(${key}).`);
+
+      // Override empty code
+      data.code = data.code || DefaultClientCode(data.client, `Feed me code at ${inst.config.codeServer}`);
+
+      // Cache the code in case the code server goes away.
+      inst.clients[key] = data;
+
+      // Forward code to clients.
+      network.emit(`code(${key})`, data);
+    });
+
+    // Handle connections from clients.
+    network.on('connection', function(socket) {
+      socket.on('requestCode', function(data) {
+
+        const key = getClientKey(data.client);
+        debug(`Client(${key}) requested code.`);
+        debug(`Code server connected: ${inst.codeServer.connected}`);
+
+        // Track the client
+        inst.clients[key] = _.extend(inst.clients[key] || {}, { client: data.client });
+        inst.clients[key] = _.defaults(inst.clients[key], { code: undefined });
+
+        let response;
+
+        if (inst.clients[key].code || !inst.codeServer.connected) {
+          debug(`Sending cached code to client(${key}).`);
+          response = _.defaults(inst.clients[key], {
+            client: data.client,
+            code: DefaultClientCode(data.client, "No code server available")
+          });
+
+          network.emit(`code(${key})`, response);
```

Comment:
This broadcasts to ALL clients. Is that what you want? Should this be socket.emit?